src/ripple/consensus/Validations.h

```diff
@@ -0,0 +1,504 @@
+        beast::uhash<>>
+        byLedger_;
+
+    //! Parameters to determine validation staleness
+    ValidationParms parms_;
+
+    //! @return The Derived class that implements the CRTP requirements.
+    Derived&
+    impl()
+    {
+        return static_cast<Derived&>(*this);
+    }
+
+protected:
+    /** Iterate current validations.
+
+        Iterate current validations, optionally removing any stale validations
+        if a time is specified
+
+        @param t (Optional) Time used to determine staleness
+        @param f Callable with signature (NodeKey const &, Validations const &)
+
+        @note This is protected since the implementations/clients primarily
+              iterate trusted validations, but the iteration logic is needed by
+              the currentKeys member function to iterate all validation keys.
+
+    */
+    template <class F>
+    void
+    current(boost::optional<NetClock::time_point> t, F&& f)
+    {
+        auto it = current_.begin();
+        while (it != current_.end())
+        {
+            // Check for staleness, if time specified
+            if (t &&
+                !isCurrent(
+                    parms_, *t, it->second.signTime(), it->second.seenTime()))
+            {
+                // contains a stale record
+                impl().onStale(std::move(it->second));
+                it = current_.erase(it);
+            }
+            else
+            {
+                auto cit = typename ValidationMap::const_iterator{it};
+                // contains a live record
+                f(cit->first, cit->second);
+                ++it;
+            }
+        }
+    }
+
+    /** Return set of validations for a ledger
+
+        Returns the set of validations for a ledger.
+        @param ledgerID The identifier of the ledger
+        @return A pointer to the set of validations, nullptr if none exist.
+    */
+    ValidationMap const*
+    byLedger(LedgerID const& ledgerID)
+    {
+        auto it = byLedger_.find(ledgerID);
+        if (it != byLedger_.end())
+        {
+            // Update set time since it is being used
+            byLedger_.touch(it);
+            return &(it->second);
+        }
+        return nullptr;
+    }
+
+    //! Return the validation timing parameters
+    ValidationParms const&
+    parms() const
+    {
+        return parms_;
+    }
+
+    //! Flush all current validations, calling `Derived::onStale` member.
+    void
+    flush()
+    {
+        for (auto& it : current_)
+        {
+            impl().onStale(std::move(it.second));
+        }
+        current_.clear();
+    }
+
+public:
+    /** Constructor
+
+        @param p ValidationParms to control staleness/expiration of validaitons
+        @param c Clock to use for expiring validations stored by ledger
+    */
+    Validations(
+        ValidationParms const& p,
+        beast::abstract_clock<std::chrono::steady_clock>& c)
+        : byLedger_(c), parms_(p)
+    {
+    }
+
```

Comment:
1. I intend `Validations` to be used as a base class via CRTP only and prefer to not have any virtual table if possible (yes premature optimization!).  I have to handle different types, so virtuals aren't much help.  Is there a technique to indicate that aside from a comment?

2. No, it must be derived from.  It think it would fail to compile otherwise due to the `static_cast` in the `impl()` helper.

I had considered avoiding CRTP altogether and just taking in a policy class that implements the `onStale` callback desired by callers.  Would that be preferable?